File: wwwroot/iCDataHandler/iCDataHandler/clsConnection.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Data;
using System.Data.SqlClient;
using MySql.Data.MySqlClient;
using iConsulting.iCXmlDbClient;

namespace iConsulting.iCDataHandler
{
	public class clsConnection
	{	
		protected string	m_sError;
		protected bool      m_HasError;
		public string	    m_sDataSource;
		public string	    m_sConnectionString;
		bool			    m_bIsConnected;
		IDbConnection	    m_oConnection;

		static string	CLASSNAME = "[Namespace::iConsulting][Service::iCDataHandler][Class::clsConnection]";

		#region Properties
		
		public string GetError
		{
			get
			{
				return this.m_sError;
			}
		}

		public bool HasError
		{
			get
			{
				return this.m_HasError; 
			}
		}

		protected string DataSource
		{
			set 
			{
				this.m_sDataSource = value;
			}
			get 
			{
				return this.m_sDataSource;
			}
		}

		protected string ConnectionString
		{
			set
			{
				this.m_sConnectionString = value;
			}
			get
			{
				return this.m_sConnectionString;
			}
		}

		protected bool IsConnected
		{
			get
			{
				return this.m_bIsConnected; 
			}
		}

		protected bool Connectable
		{
			get
			{
				if ((ValidateDataSource()) && (ValidateConnectionString()))
					return true;
				return false;
			}
		}
		#endregion

		#region Constructors

		protected clsConnection()
		{
			this.m_bIsConnected = false;
		}

		protected clsConnection(string sConnectionString)
		{
			this.m_sConnectionString = sConnectionString;
			this.m_bIsConnected = false;
		}

		protected clsConnection(string sDataSource, string sConnectionString)
		{
			this.m_sDataSource = sDataSource;
			this.m_sConnectionString = sConnectionString;
			this.m_bIsConnected = false;
		}
		
		#endregion

		#region Private Functions

		private void AddErrorData(ref string sError, Exception ex, string sFunction)
		{
			try 
			{
				sError += "#" + ex.GetType().ToString() + "occured in " + sFunction + "\r\nSource: " + ex.Source + "\r\nMessage: " + ex.Message + "\r\nVersion: " + System.Reflection.Assembly.GetExecutingAssembly().CodeBase + "\r\nCaller: " + System.Reflection.Assembly.GetCallingAssembly().CodeBase + "\r\nStack trace: " + ex.StackTrace;
				this.m_HasError = true;
				ErrorHandler(ex, sFunction);
			}
			catch (Exception)
			{
				throw;
			}
		}

		private void ErrorHandler(Exception ex, string sFunction)
		{
			try
			{
				System.Diagnostics.EventLog.WriteEntry(
					"iCDataHandler",
					ex.GetType().ToString() + "occured in " + sFunction + "\r\nSource: " + ex.Source + "\r\nMessage: " + ex.Message + "\r\nVersion: " + System.Reflection.Assembly.GetExecutingAssembly().CodeBase + "\r\nCaller: " + System.Reflection.Assembly.GetCallingAssembly().CodeBase + "\r\nStack trace: " + ex.StackTrace,
					System.Diagnostics.EventLogEntryType.Error
				);
			}
			catch (Exception)
			{
				
			}
		}

		private bool ValidateConnectionString()
		{
			string FUNCTIONNAME = CLASSNAME + "[Function::ValidateConnectionString]";
			try 
			{	        
				if(this.m_sConnectionString.Length > 0)
					return true;
				return false;
			}
			catch (Exception ex)
			{
				AddErrorData(ref this.m_sError, ex, FUNCTIONNAME);
				return false;
			}
		}

		private bool ValidateDataSource()
		{
			string FUNCTIONNAME = CLASSNAME + "[Function::ValidateDataSource]";
			try 
			{	    
				switch(this.m_sDataSource.ToLower())
				{
					case "mssqlserver":
						break;
					case "mysql":
                        break;
                    case "xml":
                        break;
					default: 
						throw new Exception(); 
				}
				return true;

			}
			catch (Exception ex)
			{
				AddErrorData(ref this.m_sError, ex, FUNCTIONNAME);
				return false;
			}
		}

		#endregion

		#region Public Functions
		
		public virtual void Initialize(string sDataSource, string sConnectionString)
		{
			this.m_sDataSource = sDataSource;
			this.m_sConnectionString = sConnectionString;
			this.m_bIsConnected = false;
		}		
	
		public IDbConnection GetConnection()
		{
			string FUNCTIONNAME = CLASSNAME + "[Function::GetConnection]";
			try
			{
				if (Connectable)
				{ 
					switch(DataSource.ToLower())
					{
						case "mssqlserver":
							this.m_oConnection = new SqlConnection(this.ConnectionString);
							this.m_oConnection.Open();
							this.m_bIsConnected = true;
							return this.m_oConnection;
						case "mysql":
							this.m_oConnection = new MySqlConnection(this.ConnectionString);
							this.m_oConnection.Open();
							this.m_bIsConnected = true;
							return this.m_oConnection;
                        case "xml":
                            this.m_oConnection = new XmlDbConnection(this.ConnectionString);
                            this.m_oConnection.Open();
                            this.m_bIsConnected = true;
                            return this.m_oConnection; 
						default:
							return this.m_oConnection;
					}
				}
				else
				{
					this.m_bIsConnected = false;
					throw new Exception();
				}			
			}
			catch (Exception ex)
			{
				AddErrorData(ref this.m_sError, ex, FUNCTIONNAME);
                this.m_bIsConnected = false;
				return this.m_oConnection;
			}
		}

		public void CloseConnection(ref IDbConnection oConnection)
		{
			string FUNCTIONNAME = CLASSNAME + "[Function::CloseConnection]";
			try
			{
				oConnection.Close();
				oConnection = null;
			}
			catch (Exception ex)
			{
				oConnection = null;
				AddErrorData(ref this.m_sError, ex, FUNCTIONNAME);
			}
			
		}

		#endregion
	}
}

File: wwwroot/iCDataManagerService/clsConnection.vb
Imports System    

Namespace iConsulting.iCMServer

    Public MustInherit Class clsConnection

        Const CLASSNAME = "[Namespace::iConsulting.iCMServer][WebService::iCDataManager][Class::clsConnection]"

        Private m_sError As String
        Private m_bHasError As Boolean
        Private m_sDataSource As String
        Private m_sConnectionString As String
        Private m_bIsConnected As Boolean

#Region " Properties "

        Public ReadOnly Property GetError() As String
            Get
                Return Me.m_sError
            End Get
        End Property

        Public Shadows ReadOnly Property HasError() As Boolean
            Get
                Return Me.m_bHasError
            End Get
        End Property

        Public Property DataSource() As String
            Get
                Return Me.m_sDataSource
            End Get
            Set(ByVal Value As String)
                Me.m_sDataSource = Value
            End Set
        End Property

        Public Property ConnectionString() As String
            Get
                Return Me.m_sConnectionString
            End Get
            Set(ByVal Value As String)
                Me.m_sConnectionString = Value
            End Set
        End Property

        Protected ReadOnly Property IsConnected() As String
            Get
                Return Me.m_bIsConnected
            End Get
        End Property

        Public ReadOnly Property Connectable() As Boolean
            Get
                If Me.ValidateDataSource And Me.ValidateConnectionString Then
                    Return True
                Else
                    Return False
                End If
            End Get
        End Property

#End Region

#Region " Constructors "

        Public Sub New()
            Me.m_bIsConnected = False
        End Sub

        Public Sub New(ByVal sConnectionString As String)
            Me.m_sConnectionString = sConnectionString
            Me.m_bIsConnected = False
        End Sub

        Public Sub New(ByVal sDataSource As String, ByVal sConnectionString As String)
            Me.m_sDataSource = sDataSource
            Me.m_sConnectionString = sConnectionString
            Me.m_bIsConnected = False
        End Sub

#End Region

#Region " Private Functions "

        Private Sub AddErrorData(ByRef sError As String, ByVal lNumber As Long, ByVal lSeverity As Long, ByVal sSource As String, ByVal sMsg As String)
            Try
                sError += "#" & lNumber & "|" & lSeverity & "|" & sSource & "|" & sMsg
                Me.m_bHasError = True
            Catch
            End Try
        End Sub

        Private Function ValidateDataSource() As Boolean
            Const FUNCTIONNAME = CLASSNAME & "[Function::ValidateDataSource]"
            Try
                Select Case Me.DataSource
                    Case ICM_DATASOURCE_MSSQLSERVER, ICM_DATASOURCE_MSACCESS, ICM_DATASOURCE_MYSQL
                        Return True
                    Case Else
                        Err.Raise(ERR_ABORT)
                End Select
            Catch
                Me.AddErrorData(Me.m_sError, Err.Number, ERR_UNKNOWN_DATASOURCE, FUNCTIONNAME, Err.Description)
                Return False
            End Try
        End Function

        Private Function ValidateConnectionString() As Boolean
            Const FUNCTIONNAME = CLASSNAME & "[Function::ValidateConnectionString]"
            Try
                If Me.m_sConnectionString = "" Then Err.Raise(ERR_ABORT)
                Return True
            Catch
                Me.AddErrorData(Me.m_sError, Err.Number, ERR_INVALID_CONNECTIONSTRING, FUNCTIONNAME, Err.Description)
                Return False
            End Try
        End Function

#End Region

#Region " Public Functions "

        Public Function GetConnection() As Object
            Const FUNCTIONNAME = CLASSNAME & "[Function::GetConnection]"
            Try
                If Me.Connectable Then
                    Select Case Me.DataSource
                        Case ICM_DATASOURCE_MSSQLSERVER
                            Dim oConnection As New SqlClient.SqlConnection(Me.ConnectionString)
                            oConnection.Open()
                            Me.m_bIsConnected = True
                            Return oConnection
                        Case ICM_DATASOURCE_MSACCESS
                            Dim oConnection As New OleDb.OleDbConnection(Me.ConnectionString)
                            oConnection.Open()
                            Me.m_bIsConnected = True
                            Return oConnection
                        Case ICM_DATASOURCE_MYSQL
                            Dim oConnection As New CoreLab.MySql.MySqlConnection(Me.ConnectionString)
                            oConnection.Open()
                            Me.m_bIsConnected = True
                            Return oConnection
                    End Select
                Else
                    Me.m_bIsConnected = False
                    Err.Raise(ERR_ABORT)
                End If
            Catch
                Me.AddErrorData(Me.m_sError, Err.Number, ERR_INVALID_CONNECTIONSTRING, FUNCTIONNAME, Err.Description)
                Me.m_bIsConnected = False
                Return New SqlClient.SqlConnection
            End Try
        End Function

        Public Sub CloseConnection(ByRef oConnection As Object)
            Const FUNCTIONNAME = CLASSNAME & "[Function::CloseConnection]"
            Try
                oConnection.Close()
                oConnection = Nothing
            Catch
                oConnection = Nothing
                Me.AddErrorData(Me.m_sError, Err.Number, ERR_CLOSE_CONNECTION, FUNCTIONNAME, Err.Description)
            End Try
        End Sub

#End Region

    End Class

End Namespace
